[fix](stream_load)fix bug when stream without content-length or chunk…

xingyingone opened the pull request:
## Proposed changes

Issue Number: close https://github.com/apache/doris/issues/27671

<!--Describe your changes.-->

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

xingyingone commented:
run buildall

HHoflittlefish777 commented on be/src/http/action/stream_load.cpp:
```diff
@@ -301,6 +301,14 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
         }
     }
 
+    if (ctx->body_bytes == 0 && !ctx->is_chunked_transfer) {
```
Ctx->body_bytes == 0 must be zero in on_header method, for it hasn't started receive data here yet, which begins in StreamLoadAction::_handle. I think you can check in  _handle or use http_req->header(HttpHeaders::CONTENT_LENGTH).empty() to check.

HHoflittlefish777 replied:
There may be several things to consider here:
1. Forgot to set content length or chunked
2. Incorrect 0 set
3. It is itself 0

HHoflittlefish777 commented on be/src/http/action/stream_load.cpp:
```diff
@@ -301,6 +301,14 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
     }
 
+    if (ctx->body_bytes == 0 && !ctx->is_chunked_transfer) {
+        LOG(WARNING) << "content_length is empty and transfer-encoding!=chunked, please set "
```
It is no need log and return InternalError is enough.

xingyingone commented:
run buildall

xingyingone commented:
run buildall

freemandealer reviewed (commented):
usually, when a function/branch is added, we need also to provide a UT or regression test. Please take a look at [here](https://doris.apache.org/zh-CN/community/developer-guide/regression-testing)

xingyingone commented:
> usually, when a function/branch is added, we need also to provide a UT or regression test. Please take a look at [here](https://doris.apache.org/zh-CN/community/developer-guide/regression-testing)

Okay , I will provide it

alexxing662 committed:
[fix](stream_load)fix bug for stream
1. forbid thed  stream_load without content-length or chunked Transfer Encoding
2. forbid thed  stream_load both with content-length and chunked Transfer Encoding

xingyingone commented:
run buildall

xingyingone commented:
run buildall

HHoflittlefish777 commented on regression-test/suites/load_p0/stream_load/test_stream_load.groovy:
```diff
@@ -1575,5 +1575,138 @@ suite("test_stream_load", "p0") {
+       def process = command.execute()
+       code = process.waitFor()
+       out = process.text
+       json2pc = parseJson(out)
```
Do not use name "json2pc", json enough

HHoflittlefish777 commented on regression-test/suites/load_p0/stream_load/test_stream_load.groovy:
```diff
@@ -1575,5 +1575,138 @@ suite("test_stream_load", "p0") {
+        sql """ DROP TABLE IF EXISTS ${tableName16} FORCE"""
+    }
+
+
```
Can you add test with empty file?

xingyingone commented:
run buildall

xingyingone commented:
run buildall

xingyingone commented:
run buildall

sollhui reviewed (approved):
LGTM

xingyingone commented:
run buildall

freemandealer commented on be/src/http/action/stream_load.cpp:
```diff
@@ -302,12 +309,21 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
             ctx->is_chunked_transfer = true;
         }
     }
+    if (http_req->header(HttpHeaders::CONTENT_LENGTH).empty() && !ctx->is_chunked_transfer) {
```
please add UNLIKELY to make branch predictor happy

freemandealer commented on be/src/http/action/stream_load.cpp:
```diff
@@ -302,12 +309,21 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
+        return Status::InternalError(
+                "content_length is empty and transfer-encoding!=chunked, please set content_length "
+                "or transfer-encoding=chunked");
+    } else if (!http_req->header(HttpHeaders::CONTENT_LENGTH).empty() && ctx->is_chunked_transfer) {
```
ditto, UNLIKELY

freemandealer commented on regression-test/suites/load_p0/stream_load/test_stream_load.groovy:
```diff
@@ -1575,5 +1575,136 @@ suite("test_stream_load", "p0") {
+        out = process.text
+        log.info("test chunked transfer result: ${out}".toString())
+        def json = parseJson(out)
+        assertEquals("success", json.Status.toLowerCase())
```
why success?

freemandealer commented on regression-test/suites/load_p0/stream_load/test_stream_load.groovy:
```diff
@@ -1575,5 +1575,136 @@ suite("test_stream_load", "p0") {
+        out = process.text
+        log.info("test chunked transfer result: ${out}".toString())
+        def json = parseJson(out)
+        assertEquals("success", json.Status.toLowerCase())
```
why success?

freemandealer commented on be/src/http/action/stream_load.cpp:
```diff
@@ -144,6 +144,13 @@ Status StreamLoadAction::_handle(std::shared_ptr<StreamLoadContext> ctx) {
         LOG(WARNING) << "recevie body don't equal with body bytes, body_bytes=" << ctx->body_bytes
                      << ", receive_bytes=" << ctx->receive_bytes << ", id=" << ctx->id;
         return Status::InternalError("receive body don't equal with body bytes");
+    } else if (ctx->body_bytes == 0 && ctx->receive_bytes && ctx->is_chunked_transfer == false) {
```
unlikely

xingyingone commented:
run buildall

xingyingone commented:
run buildall

xingyingone commented:
run buildall

dataroaring reviewed (approved):
LGTM
